### PR TITLE
[quant] Quantizable LSTM

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49866 [quant] Quantizable MultiheadAttention
* **#49671 [quant] Quantizable LSTM**

- Introduces the `torch.nn.quantizable` namespace
- Adds the `torch.nn.quantizable.LSTM` module

The point of the `quantizable` namespace is to segregate the purely quantized modules with the modules that could be quantized through a normal quantization flow, but are not using the quantized kernels explicitly.
That means the quantizable modules are functionally and numerically equivalent to the FP ones and can be used instead of the FP ones without any loss.

The main difference between the `torch.nn.LSTM` and the `torch.nn.quantizable.LSTM` is that the former one does not support observation for the linear layers, because all the computation is internal to the `aten` namespace.
The `torch.nn.quantizable.LSTM`, however, uses explicit linear layers that can be observed for further quantization.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_custom_module_lstm
```

Differential Revision: [D25663870](https://our.internmc.facebook.com/intern/diff/D25663870)

### CODE REVIEW

**Comment by z-a-f** [comment]
The original was rewritten, but the comments are still valid: https://github.com/pytorch/pytorch/pull/46229

**Comment by z-a-f** [comment]
@zrphercule Can you try out this PR and see if you get an error?

**Comment by zrphercule** [comment]
No hay problema!

Zafar <notifications@github.com>于2020年12月21日 周一上午2:54写道：

File: test/quantization/test_quantized_op.py [hunk 1]
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
                 self.assertEqual(result_ref[0], result_dynamic[0], msg="torch.quantized_rnncell results are off")
 
 
+"""Tests recurrent layers."""
+class TestPTQRNN(TestCase):
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:2738]: can you merge this test case with TestQuantizedOps: https://github.com/pytorch/pytorch/blob/master/test/quantization/test_quantized_op.py#L124

File: test/test_quantization.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 from quantization.test_quantized_op import TestComparatorOps  # noqa: F401
 from quantization.test_quantized_op import TestPadding  # noqa: F401
 from quantization.test_quantized_op import TestQuantizedEmbeddingOps  # noqa: F401
```
Comment (jerryzh168) [inline test/test_quantization.py:16]: I think we should merge these with TestQuantizedOps as well

File: torch/__init__.py [hunk 3]
```diff
@@ -574,6 +574,7 @@ def _assert(condition, message):
 import torch.futures
 import torch.nn
 import torch.nn.intrinsic
+import torch.nn.quantizable
```
Comment (jerryzh168) [inline torch/__init__.py:577]: should we put this under torch.nn.mo.quantizable since we are planning to move everything under mo?

File: test/quantization/test_quantized_op.py [hunk 4]
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
...
+        custom_module_config = {
+            'float_to_observed_custom_module_class': {
+                torch.nn.LSTM: torch.nn.quantizable.LSTM
+            }
+        }
+
+        # Moving the mean of the random input to avoid signal being close to 0
+        x = np.random.randn(seq_len, batch_size, input_size)
+        scale, zero_point = _calculate_dynamic_qparams(x, dtype=dtype)
+        qx = _quantize(x, scale=scale, zero_point=zero_point, dtype=dtype)
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:2785]: this is overwritten by L2787, do we need this?

File: test/quantization/test_quantized_op.py [hunk 5]
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
...
+        dropout = 0  # This is not supported
+
+        Bias = [False, True]
+        Batch_first = [False, True]
+        Bidirectional = [False, True]
+
+        dtype = np.uint8
+        qtype = torch.quint8
+
+        custom_module_config = {
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:2776]: should we add this support in the eager mode and fx graph mode flow by default instead of requiring user to do this every time?

File: test/quantization/test_quantized_op.py [hunk 6]
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
...
+                    lstm, prepare_custom_config_dict=custom_module_config)
+                self.assertTrue(hasattr(lstm_prepared[0], 'layers'))
+                self.assertEqual(num_layers, len(lstm_prepared[0].layers))
+
+                # Calibrate
+                y = lstm_prepared(x)
+                self.assertEqual(y_ref, y)
+
+                # Quantize
+                lstm_quantized = torch.quantization.convert(lstm_prepared)
```
Comment (jerryzh168) [inline test/quantization/test_quantized_op.py:2827]: also, how is convert supported? I didn't see quantized lstm module, also there is no convert_custom_config_dict here.

File: torch/__init__.py (hunk 3 above)
Comment (z-a-f) [reply torch/__init__.py:577]: I was thinking about it -- but currently `mo` is in the `fb` namespace. Should I start introducing the new namespace?

File: test/quantization/test_quantized_op.py (hunk 4 above)
Comment (z-a-f) [reply test/quantization/test_quantized_op.py:2785]: Oh, my bad!

File: test/quantization/test_quantized_op.py (hunk 6 above)
Comment (z-a-f) [reply test/quantization/test_quantized_op.py:2827]: Currently, the convert treats the lstm as if it was just another module -- as long as it is quantizable, all of its layers are just replaced with the quantized ones. We cannot do it for prepare, because we need to avoid the insertion of the observers.

File: torch/nn/quantizable/modules/rnn.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+"""
+We will recreate all the RNN modules as we require the modules to be decomposed
+into its building blocks to be able to observe.
+"""
+
+class LSTMCell(torch.nn.Module):
+    _FLOAT_MODULE = torch.nn.LSTMCell
+    # There are multiple outputs in the forward -- we don't want to observe!
+    do_not_observe: bool = True
```
[the remark this answers is not in the document]
Comment (z-a-f) [reply torch/nn/quantizable/modules/rnn.py:16]: It's introduced in the previous PR

File: test/quantization/test_quantized_op.py (hunk 5 above)
Comment (raghuramank100) [reply test/quantization/test_quantized_op.py:2776]: This is not done by default. For default users, we do module swapping and called the fused quantized lstm op. This is done only for cases when users need a lowered IR that contains the underlying linear ops.

File: test/quantization/test_quantized_op.py [hunk 8]
```diff
@@ -2734,6 +2734,109 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
...
+        dtype = np.uint8
+        qtype = torch.quint8
+
+        custom_module_config = {
+            'float_to_observed_custom_module_class': {
+                torch.nn.LSTM: torch.nn.quantizable.LSTM
+            }
+        }
+
+        # Moving the mean of the random input to avoid signal being close to 0
```
Comment (raghuramank100) [inline test/quantization/test_quantized_op.py:2782]: Is this comment applicable now? The input is zero mean.

File: torch/nn/quantizable/modules/rnn.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+import torch
+from torch import Tensor
+
+"""
+We will recreate all the RNN modules as we require the modules to be decomposed
+into its building blocks to be able to observe.
+"""
+
+class LSTMCell(torch.nn.Module):
```
Comment (vkuzo) [inline torch/nn/quantizable/modules/rnn.py:13]: for all of these, can we add a `_get_name` method which includes the information that this is not the vanilla `nn.LSTMCell`, etc

File: torch/nn/quantizable/modules/rnn.py [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+        cell.igates.weight = torch.nn.Parameter(wi)
+        if bi is not None:
+            cell.igates.bias = torch.nn.Parameter(bi)
+        cell.hgates.weight = torch.nn.Parameter(wh)
+        if bh is not None:
+            cell.hgates.bias = torch.nn.Parameter(bh)
+        return cell
+
+    @classmethod
+    def from_float(cls, other):
```
Comment (vkuzo) [inline torch/nn/quantizable/modules/rnn.py:80]: IMO it would be cleaner if this class had 0 references to quantization, and things like attaching the qconfig, etc, would be done by the user outside of this class.  This would remove the dependency of this module on quantization and make the code more modular.  What are your thoughts?

File: torch/testing/_internal/common_quantized.py [hunk 11]
```diff
@@ -102,6 +102,27 @@ def _calculate_dynamic_per_channel_qparams(X, dtype):
 
     return scale, zero_point
 
+def _snr(x, x_hat):
```
Comment (vkuzo) [inline torch/testing/_internal/common_quantized.py:105]: nit: can we annotate types on function I/O

**Review by vkuzo — COMMENTED** [review]
makes sense, had some questions inline.  High level, can you comment on what are the performance implications of this in fp32, and why not just change the existing LSTM module.  I remember we talked about this some time back, but would be great to document the tradeoff here.

File: torch/nn/quantizable/modules/rnn.py (hunk 10 above)
Comment (z-a-f) [reply torch/nn/quantizable/modules/rnn.py:80]: That violates the whole quantization flow as we have it. The way we use both quantized and quantizable modules (custom or otherwise) is that we call the `from_float` on the FP module, which is expected to have the qconfig -- if it does not, that means the FP module should not be quantized. Also, the quantizable modules are not full replacement of the floating point ones -- they are designed to be used the same way the quantized ones are. The reason they are in a different namespace (as per discussion) is that this namespace does not have any quantized kernels, and can be run on the FP tensors directly.

File: torch/testing/_internal/common_quantized.py (hunk 11 above)
Comment (z-a-f) [reply torch/testing/_internal/common_quantized.py:105]: This is a very good point. The short answer is **maybe**, but I don't see the reason why :)

The long answer is that the files under the `testing/_internal` are not expected to get jitted or to have a mandatory type checking. This means that other than readability the type hints in this folder will serve no purpose.

However, because mot functions under testing can handle multiple argument types, the readability will suffer a lot if type hints are added. For example, here how this function will look like:

```python
from typing import List, Tuple
from torch import Tensor

ListOfTensors = List[Union[List, Tuple, Tensor]]
TupleOfTensors = Tuple[Union[List, Tuple, Tensor]]

def _snr(x: Union[Tensor, ListOfTensors, TupleOfTensors],
         x_hat: Union[Tensor, ListOfTensors, TupleOfTensors]
) -> Union[Tensor, ListOfTensors, TupleOfTensors]:
    ...

# Or, if we are OK with being less descriptive:

def _snr(x: Union[Tensor, List, Tuple],
         x_hat: Union[Tensor, List, Tuple]
) -> Union[Tensor, List, Tuple]:
    ...

```

If you notice, the readability of the code does not improve, and because the function is not expected to get type checked, there are only downsides of annotating it. With that said, if we decide to annotate the rest of the code under the `testing/_internal`, I think we can start now

Comment (vkuzo) [reply torch/testing/_internal/common_quantized.py:105]: > However, because mot functions under testing can handle multiple argument types, the readability will suffer a lot if type hints are added. For example, here how this function will look like:

The answer might vary per person, but having types significantly improves readability for me personally, even if it might decrease it for you.  Are there any compromises we can go for here?

File: torch/nn/quantizable/modules/rnn.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+                       other.bidirectional)
+        observed.qconfig = getattr(other, 'qconfig', qconfig)
+        for idx in range(other.num_layers):
+            observed.layers[idx] = _LSTMLayer.from_float(other, idx, qconfig,
+                                                         batch_first=False)
+        observed.eval()
+        observed = torch.quantization.prepare(observed, inplace=True)
+        return observed
+
+    def from_observed(self, other):
```
Comment (vkuzo) [inline torch/nn/quantizable/modules/rnn.py:292]: I see the calls to this in the custom module code, but this PR seems like the first non-test use case.  Is that expected?  Should this be an interface with a docblock somewhere?

If this is the only callsite, should we even have this API at all?  Can we just inline the `torch.quantization.convert` call in the custom module flow instead, and have one less API surface to worry about?  cc @jerryzh168

File: torch/nn/quantizable/modules/rnn.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+
+"""
+We will recreate all the RNN modules as we require the modules to be decomposed
+into its building blocks to be able to observe.
+"""
+
+class LSTMCell(torch.nn.Module):
+    _FLOAT_MODULE = torch.nn.LSTMCell
+
+    def __init__(self, input_dim, hidden_dim, bias=True):
```
Comment (vkuzo) [inline torch/nn/quantizable/modules/rnn.py:16]: can we make sure mypy is enabled for this module (in caffe2/mypy.ini), and annotate types on at least function I/O?   It would help with readability.

File: torch/nn/quantizable/modules/rnn.py (hunk 10 above)
Comment (vkuzo) [reply torch/nn/quantizable/modules/rnn.py:83]: sg.  I think the `quantizable` naming is weird to think about, but makes sense, thanks.

File: torch/nn/quantizable/modules/rnn.py (hunk 12 above)
Comment (z-a-f) [reply torch/nn/quantizable/modules/rnn.py:292]: We had a minor discussion with @jerry39213gh about exactly that -- the reason we want to have the `from_observed` is because the conversion process is not necessarily producing structurally equivalent module. For example, if we decide to rewrite the LSTM in C++, and have a module that just calls it in the `forward`, we could. In such a case, the `from_float` would convert the FP module into the "quantizable" version that will be calibrated/observed, and later converted using the `from_observed`. In such (hypothetical) scenario, the `from_observed` would not call the `convert`, but explicitly remove the internal linear layers and replace all of them with a single call to an `aten` kernel.

As for the docs: This will be documented the same way the `from_float` is documented: Once the custom module is more or less stable, we will add a tutorial + a paragraph in the quantization documentation

File: torch/testing/_internal/common_quantized.py (hunk 11 above)
Comment (z-a-f) [reply torch/testing/_internal/common_quantized.py:105]: Wait, I thought I already added args description in the code to avoid the type annotations. Can you, please, refresh?

File: torch/nn/quantizable/modules/rnn.py (hunk 13 above)
Comment (z-a-f) [reply torch/nn/quantizable/modules/rnn.py:16]: SG!

File: torch/testing/_internal/common_quantized.py (hunk 11 above)
Comment (vkuzo) [reply torch/testing/_internal/common_quantized.py:105]:
```
ListOfTensors = List[Union[List, Tuple, Tensor]]
TupleOfTensors = Tuple[Union[List, Tuple, Tensor]]

def _snr(x: Union[Tensor, ListOfTensors, TupleOfTensors],
         x_hat: Union[Tensor, ListOfTensors, TupleOfTensors]
) -> Union[Tensor, ListOfTensors, TupleOfTensors]:
```

if you're up for doing this, that would be awesome.  I think it's much more readable than not having it.

**Review by vkuzo — APPROVED** [review]
